src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/iceberg/IcebergSchemaEvolutionService.java

```diff
@@ -17,39 +20,85 @@ public class IcebergSchemaEvolutionService implements SchemaEvolutionService {
-    if (!columnsToAdd.isEmpty()) {
-      LOGGER.debug("Adding columns to iceberg table: {} columns: {}", tableName, columnsToAdd);
-      TableSchema tableSchema =
-          tableSchemaResolver.resolveTableSchemaFromRecord(record, columnsToAdd);
+    if (!columnsToEvolve.isEmpty()) {
+      LOGGER.debug("Adding columns to iceberg table: {} columns: {}", tableName, columnsToEvolve);
+      // some of the column might already exist, and we will modify them, not create
+      List<IcebergColumnTree> alreadyExistingColumns =
+          icebergTableSchemaResolver.resolveIcebergSchemaFromChannel(
+              schemaAlreadyInUse, columnsToEvolve);
+
+      // new columns resolved from incoming record
+      List<IcebergColumnTree> modifiedOrAddedColumns =
+          icebergTableSchemaResolver.resolveIcebergSchemaFromRecord(record, columnsToEvolve);
+
+      // columns that we simply add because they are NOT present in an already existing schema.
+      List<IcebergColumnTree> addedColumns =
+          modifiedOrAddedColumns.stream()
+              .filter(
+                  modifiedOrAddedColumn ->
+                      alreadyExistingColumns.stream()
+                          .noneMatch(
+                              tree ->
+                                  tree.getColumnName()
+                                      .equalsIgnoreCase(modifiedOrAddedColumn.getColumnName())))
+              .collect(Collectors.toList());
+      // column that are present in a schema and needs to have its type modified
+      List<IcebergColumnTree> modifiedColumns =
+          modifiedOrAddedColumns.stream()
+              .filter(
+                  modifiedOrAddedColumn ->
+                      alreadyExistingColumns.stream()
+                          .anyMatch(
+                              tree ->
+                                  tree.getColumnName()
+                                      .equalsIgnoreCase(modifiedOrAddedColumn.getColumnName())))
+              .collect(Collectors.toList());
+
+      Optional<String> addColumnsQuery = generateAddColumnQuery(addedColumns);
+
+      // merge changes into already existing column
+      alreadyExistingColumns.forEach(
+          existingColumn -> {
+            IcebergColumnTree mewVersion =
+                modifiedColumns.stream()
+                    .filter(c -> c.getColumnName().equals(existingColumn.getColumnName()))
+                    .collect(Collectors.toList())
+                    .get(0);
+            existingColumn.merge(mewVersion);
+          });
+      Optional<String> alterSetDataTypeQuery = alterSetDataTypeQuery(alreadyExistingColumns);
       try {
-        conn.appendColumnsToIcebergTable(tableName, tableSchema.getColumnInfos());
+        addColumnsQuery.ifPresent(query -> conn.evolveIcebergColumns(tableName, query));
+        alterSetDataTypeQuery.ifPresent(query -> conn.evolveIcebergColumns(tableName, query));
```

Comment:
I would split this big wall of text to at least two methods: addingColumns, alteringColumnTypes